In [1]:
import pandas as pd
import numpy as np
import json
import glob
from sklearn import preprocessing

In [2]:
# Cargamos todos los archivos json que existen en la ruta seleccionada
path = r'C:\Users\Equipo\Desktop\BIG DATA\TFM\WEMOB'
all_files = glob.glob(path + "\*.json")
li = []
for filename in all_files:
    df = pd.read_json(filename, lines = True)
    li.append(df)

wemobDF = pd.concat(li)

In [3]:
wemobDF.shape

(10729032, 21)

In [4]:
df = wemobDF.copy()

In [5]:
# creamos columna id, ya que más adelante voy a separar timestamp para hacer imputaciones y eliminar NAs
# para después unir de nuevo por id, y poder mantener variable Timestamp
df['Id'] = range(1, 10729033, 1)

In [6]:
df

Latitud  Longitud  Altitud                 Timestamp    Compania  \
0       37.390165 -5.938611       21 2020-01-21 23:11:22+00:00  ALIAGUILLA   
1       37.390612 -5.938944       21 2020-01-21 23:12:22+00:00  ALIAGUILLA   
2       37.391055 -5.939556       21 2020-01-21 23:13:22+00:00  ALIAGUILLA   
3       37.391029 -5.939750       21 2020-01-21 23:14:22+00:00  ALIAGUILLA   
4       37.390972 -5.939639       21 2020-01-22 02:55:01+00:00  ALIAGUILLA   
...           ...       ...      ...                       ...         ...   
507803  39.502860 -2.269250      803 2020-06-22 21:55:46+00:00  ALIAGUILLA   
507804  39.508390 -2.253667      802 2020-06-22 21:56:46+00:00  ALIAGUILLA   
507805  39.513611 -2.238083      831 2020-06-22 21:57:46+00:00  ALIAGUILLA   
507806  39.516888 -2.224333      874 2020-06-22 21:58:46+00:00  ALIAGUILLA   
507807  39.509418 -2.211278      879 2020-06-22 21:59:46+00:00  ALIAGUILLA   

                  Matricula      Flota                 Conductor    Odometro  \
0                   3306JZR  TRACTORAS                        --  609295.105   
1                   3306JZR  TRACTORAS                        --  609295.185   
2                   3306JZR  TRACTORAS                        --  609295.295   
3                   3306JZR  TRACTORAS                        --  609295.365   
4                   3306JZR  TRACTORAS                        --  609295.375   
...                     ...        ...                       ...         ...   
507803  5497LFN ITER FRUGAL  TRACTORAS  UNKNOWN E28672116V000000   79639.815   
507804  5497LFN ITER FRUGAL  TRACTORAS  UNKNOWN E28672116V000000   79641.310   
507805  5497LFN ITER FRUGAL  TRACTORAS  UNKNOWN E28672116V000000   79642.800   
507806  5497LFN ITER FRUGAL  TRACTORAS  UNKNOWN E28672116V000000   79644.095   
507807  5497LFN ITER FRUGAL  TRACTORAS  UNKNOWN E28672116V000000   79645.550   

        Velocidad  ...  TemperaturaMotor      RPM  NivelFuel  ConsumoTotal  \
0           0.000  ...               0.0     0.00       70.8      192040.5   
1          24.472  ...              42.0   887.00       70.8      192040.5   
2           4.812  ...              44.0   620.25       70.8      192040.5   
3           3.376  ...              46.0   688.00       70.8      192041.0   
4           0.000  ...               0.0     0.00       70.8      192041.0   
...           ...  ...               ...      ...        ...           ...   
507803     91.748  ...              78.0  1275.00       49.6       25233.0   
507804     92.160  ...              78.0  1278.00       49.2       25233.5   
507805     87.040  ...              79.0  1210.00       49.2       25234.5   
507806     77.616  ...              84.0  1388.00       49.2       25235.5   
507807     92.160  ...              86.0  1279.00       49.2       25236.5   

        Frenadas  TiempoRalenti   ParMotor     Pedal  Embrague        Id  
0         5807.0      5847793.0  3465938.0  123438.0       NaN         1  
1         5807.0      5847816.0  3465938.0  123441.0       NaN         2  
2         5807.0      5847841.0  3465938.0  123441.0       NaN         3  
3         5807.0      5847861.0  3465938.0  123444.0       NaN         4  
4         5807.0      5847890.0  3465938.0  123444.0       NaN         5  
...          ...            ...        ...       ...       ...       ...  
507803     251.0       589429.0   373818.0    6983.0   59592.0  10729028  
507804     251.0       589429.0   373818.0    6983.0   59592.0  10729029  
507805     251.0       589429.0   373831.0    6983.0   59592.0  10729030  
507806     251.0       589429.0   373890.0    6983.0   59593.0  10729031  
507807     251.0       589429.0   373915.0    6983.0   59594.0  10729032  

[10729032 rows x 22 columns]

In [7]:
df.dtypes

Latitud                         float64
Longitud                        float64
Altitud                           int64
Timestamp           datetime64[ns, UTC]
Compania                         object
Matricula                        object
Flota                            object
Conductor                        object
Odometro                        float64
Velocidad                       float64
HorasMotor                      float64
ControlCrucero                  float64
TemperaturaMotor                float64
RPM                             float64
NivelFuel                       float64
ConsumoTotal                    float64
Frenadas                        float64
TiempoRalenti                   float64
ParMotor                        float64
Pedal                           float64
Embrague                        float64
Id                                int32
dtype: object

In [8]:
df.isnull().sum()

Latitud                   0
Longitud                  0
Altitud                   0
Timestamp                 0
Compania                  0
Matricula                 0
Flota                     0
Conductor                 0
Odometro               2521
Velocidad                63
HorasMotor               70
ControlCrucero         1288
TemperaturaMotor      22466
RPM                   21697
NivelFuel             33792
ConsumoTotal         482498
Frenadas             367161
TiempoRalenti           855
ParMotor               1538
Pedal                   255
Embrague            1883566
Id                        0
dtype: int64

In [9]:
df['Compania'].unique()

array(['ALIAGUILLA'], dtype=object)

In [10]:
df['Flota'].unique()

array(['TRACTORAS'], dtype=object)

In [11]:
# Guardamos en una lista los conductores
listaConductores = df['Conductor'].unique()
listaConductores

array(['--', 'Francisco Herrera Molina', 'Antonio Rodriguez Cabezas',
       'Manuel Moreno Oropesa', 'Francisco Javier Herrera',
       'UNKNOWN E42091266R000002', 'Jose Antonio Sanchez Lopez',
       'Fernando Alfredo Guevara Benitez', 'Rosendo Gonzalez Hermida',
       'Juan Jose Francisco Reina', 'UNKNOWN E44603983Z000000',
       'Vinicius Dantas Barbosa', 'UNKNOWN E45651082S000002',
       'Antonio Librero Rubio', 'UNKNOWN E28822209N000002',
       'UNKNOWN EX5668726P000000', 'Javier Cacho',
       'UNKNOWN E71281425D000000', 'Krasimir Tsenov Ivanov',
       'Ancizar Cardona Salazar', 'Marcelino', 'Luis Vidal',
       'UNKNOWN E71290702V000010', 'Ruben Velasco', 'M.Angel Poveda',
       'UNKNOWN E44493612C000000', 'UNKNOWN E71522135R000000',
       'UNKNOWN E71449800R000001', 'Eduardo Garcia Gonzalez',
       'UNKNOWN E44430028P000002', 'Jose Maria Herrera Molina',
       'UNKNOWN E53145122B000002', 'Daniel Todorov Atanasov',
       'Jose Manuel Moreno Garcia', 'Jose Carlos Valca

In [12]:
# Contamos los conductores, restamos uno por los desconocidos que aparecen como --
print("Existen un total de " + str(len(listaConductores)-1) + " conductores")

Existen un total de 292 conductores


Nos interesan los conductores con valor '--', para eliminarlos por ser desconocidos, los pasamos a Nan y luego eliminamos. 

In [13]:
df['Conductor'].replace('--', np.nan, inplace=True)

In [14]:
df = df.dropna(axis=0, subset=['Conductor'])

In [15]:
df.shape

(10323170, 22)

In [16]:
# Vemos como quedaría la conversión de conductores
le = preprocessing.LabelEncoder()
le.fit(df['Conductor'])
le_name_mapping = dict(zip(le.classes_, le.transform(le.classes_)))
print(le_name_mapping)

{'Abdellatif Chaairi Fatah': 0, 'Abel De Paz Hidalgo': 1, 'Agustin Lopez Lopez': 2, 'Alejandro Gonzalez': 3, 'Ancizar Cardona Salazar': 4, 'Angel Santos Ferrero Garcia': 5, 'Antonio Gordon Gomez': 6, 'Antonio Librero Rubio': 7, 'Antonio Lopez Villarejo': 8, 'Antonio Martinez Bravo': 9, 'Antonio Olmos Lizan': 10, 'Antonio Rodriguez Cabezas': 11, 'Arley Vasquez Valencia': 12, 'Constantin Dan Antonia': 13, 'Daniel Todorov Atanasov': 14, 'David Gomez Marcos': 15, 'David Laup Cañizares': 16, 'Dinitar': 17, 'Domingo Pastrian': 18, 'Eduardo Garcia Gonzalez': 19, 'Eduardo Rodriguez Saavedra': 20, 'Emilio Perez Gonzalez': 21, 'Emilio Remesal Guzman': 22, 'Fernando Alfredo Guevara Benitez': 23, 'Fernando Martinez Sanchez': 24, 'Florencio Santos Carretero': 25, 'Francisco Herrera Molina': 26, 'Francisco Javier Alvarez Lobo': 27, 'Francisco Javier Cortinas': 28, 'Francisco Javier Herrera': 29, 'Francisco Javier Rusillo Camara': 30, 'Francisco Mariñas Lopez': 31, 'Francisco Rodriguez Marin': 32, 'F

In [17]:
from sklearn.preprocessing import LabelEncoder

In [18]:
## Encode target labels with value between 0 and n_classes-1., lo transforma en frecuencia
transform1 = LabelEncoder()
transform1.fit_transform(listaConductores)
df['ConductorNum'] = transform1.transform(df['Conductor'])

C:\Users\Equipo\anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


In [19]:
df.dtypes

Latitud                         float64
Longitud                        float64
Altitud                           int64
Timestamp           datetime64[ns, UTC]
Compania                         object
Matricula                        object
Flota                            object
Conductor                        object
Odometro                        float64
Velocidad                       float64
HorasMotor                      float64
ControlCrucero                  float64
TemperaturaMotor                float64
RPM                             float64
NivelFuel                       float64
ConsumoTotal                    float64
Frenadas                        float64
TiempoRalenti                   float64
ParMotor                        float64
Pedal                           float64
Embrague                        float64
Id                                int32
ConductorNum                      int32
dtype: object

In [20]:
df.head()

Latitud  Longitud  Altitud                 Timestamp    Compania  \
6   37.390972 -5.939639       21 2020-01-22 02:57:01+00:00  ALIAGUILLA   
7   37.390972 -5.939639       21 2020-01-22 02:58:01+00:00  ALIAGUILLA   
8   37.390972 -5.939639       21 2020-01-22 02:59:01+00:00  ALIAGUILLA   
9   37.390972 -5.939639       21 2020-01-22 03:00:01+00:00  ALIAGUILLA   
10  37.390972 -5.939639       21 2020-01-22 03:01:01+00:00  ALIAGUILLA   

   Matricula      Flota                 Conductor    Odometro  Velocidad  ...  \
6    3306JZR  TRACTORAS  Francisco Herrera Molina  609295.375        0.0  ...   
7    3306JZR  TRACTORAS  Francisco Herrera Molina  609295.375        0.0  ...   
8    3306JZR  TRACTORAS  Francisco Herrera Molina  609295.375        0.0  ...   
9    3306JZR  TRACTORAS  Francisco Herrera Molina  609295.375        0.0  ...   
10   3306JZR  TRACTORAS  Francisco Herrera Molina  609295.375        0.0  ...   

       RPM  NivelFuel  ConsumoTotal  Frenadas  TiempoRalenti   ParMotor  \
6   549.75       70.4      192041.0    5807.0      5847999.0  3465938.0   
7   549.25       70.4      192041.0    5807.0      5848059.0  3465938.0   
8   549.00       70.4      192041.0    5807.0      5848119.0  3465938.0   
9   550.00       70.4      192041.0    5807.0      5848179.0  3465938.0   
10  548.75       70.4      192041.0    5807.0      5848239.0  3465938.0   

       Pedal  Embrague  Id  ConductorNum  
6   123444.0       NaN   7            27  
7   123444.0       NaN   8            27  
8   123444.0       NaN   9            27  
9   123444.0       NaN  10            27  
10  123444.0       NaN  11            27  

[5 rows x 23 columns]

In [21]:
# Vemos como quedaría la conversión (esto hay que verificarlo) de matriculas, salen 112 matrículas
le = preprocessing.LabelEncoder()
le.fit(df['Matricula'])
le_name_mapping = dict(zip(le.classes_, le.transform(le.classes_)))
print(le_name_mapping)

{'\t7663JYK': 0, '0254JFX': 1, '0257JFX': 2, '0806JYG': 3, '1222JFR': 4, '1371KNV': 5, '1372KNV': 6, '1382JKN': 7, '1395KNV': 8, '1413KNV': 9, '1505KJF': 10, '1510KJF': 11, '1511KJF': 12, '1628KKC': 13, '1632KKC': 14, '1635KKC': 15, '1636KKC': 16, '1637KKC': 17, '1834HVB': 18, '2187KMV': 19, '2188KMV': 20, '2193KMV': 21, '2195JJJ': 22, '2195KMV': 23, '2214KMV': 24, '2228KLR': 25, '2246KLR': 26, '2351KLR': 27, '2568HTX': 28, '2574HTX': 29, '2606KRC': 30, '2607KRC': 31, '2737JXG': 32, '2739JXG': 33, '2861JBT': 34, '2868HWN': 35, '3067JBD': 36, '3214JTT': 37, '3301JZR': 38, '3303JZR': 39, '3304JZR': 40, '3306JZR': 41, '3307JZR': 42, '3308JZR': 43, '3534KGB': 44, '3536KGB': 45, '3537KGB': 46, '3893HCX': 47, '3896JBL': 48, '3897JBL': 49, '4053KWY': 50, '4054KWY': 51, '4056KWY': 52, '4058KWY': 53, '4059KWY': 54, '4060KWY': 55, '4061KWY': 56, '4062KWY': 57, '4063KWY': 58, '4065KWY': 59, '4391JLC': 60, '4774KTM': 61, '4838LGF': 62, '5497LFN ITER FRUGAL': 63, '5498LFN ITER FRUGAL': 64, '5525LBN

In [22]:
# Guardamos en una lista las matrículas
listaMatriculas = df['Matricula'].unique()
listaMatriculas

array(['3306JZR', '3303JZR', '8675JFB', '3536KGB', '1834HVB', '3537KGB',
       '7170LDY', '7066JWW', '3534KGB', '7069JWW', '0257JFX', '8900KGW',
       '8898KGW', '2195JJJ', '2568HTX', '0806JYG', '7133JST', '8599KSD',
       '7657JYK', '\t7663JYK', '2861JBT', '4774KTM', '2574HTX', '8909KGW',
       '3897JBL', '9311JFL', '2607KRC', '0254JFX', '3307JZR', '9357JWP',
       '3301JZR', '9530JFP', '8134JXN', '9278HWC', '2214KMV', '9363JWP',
       '1510KJF', '9238KGK', '4391JLC', '3304JZR', '8592KSD', '1505KJF',
       '1635KKC', '1511KJF', '1222JFR', '1636KKC', '8579HRZ', '6002KLH',
       '2228KLR', '6061KLH', '2351KLR', '7438JYD', '9269HWC', '1382JKN',
       '2188KMV', '2193KMV', '2195KMV', '2246KLR', '2187KMV', '7441JYD',
       '1413KNV', '3067JBD', '1395KNV', '1372KNV', '7482KPF',
       '7594KPF-BAJA', '7500KPF', '7480KPF', '8133JXN', '1637KKC',
       '2606KRC', '8810KRZ', '8980KRZ', '8445KRZ', '8429KRZ', '8806KRZ',
       '8496KRZ', '8440KRZ', '8454KRZ', '8591KSD', '8598KSD', '859

In [23]:
## Encode target labels with value between 0 and n_classes-1., lo transforma en frecuencia
transform2 = LabelEncoder()
transform2.fit_transform(listaMatriculas)
df['MatriculaNum'] = transform2.transform(df['Matricula'])

C:\Users\Equipo\anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


In [24]:
df

Latitud  Longitud  Altitud                 Timestamp    Compania  \
6       37.390972 -5.939639       21 2020-01-22 02:57:01+00:00  ALIAGUILLA   
7       37.390972 -5.939639       21 2020-01-22 02:58:01+00:00  ALIAGUILLA   
8       37.390972 -5.939639       21 2020-01-22 02:59:01+00:00  ALIAGUILLA   
9       37.390972 -5.939639       21 2020-01-22 03:00:01+00:00  ALIAGUILLA   
10      37.390972 -5.939639       21 2020-01-22 03:01:01+00:00  ALIAGUILLA   
...           ...       ...      ...                       ...         ...   
507803  39.502860 -2.269250      803 2020-06-22 21:55:46+00:00  ALIAGUILLA   
507804  39.508390 -2.253667      802 2020-06-22 21:56:46+00:00  ALIAGUILLA   
507805  39.513611 -2.238083      831 2020-06-22 21:57:46+00:00  ALIAGUILLA   
507806  39.516888 -2.224333      874 2020-06-22 21:58:46+00:00  ALIAGUILLA   
507807  39.509418 -2.211278      879 2020-06-22 21:59:46+00:00  ALIAGUILLA   

                  Matricula      Flota                 Conductor    Odometro  \
6                   3306JZR  TRACTORAS  Francisco Herrera Molina  609295.375   
7                   3306JZR  TRACTORAS  Francisco Herrera Molina  609295.375   
8                   3306JZR  TRACTORAS  Francisco Herrera Molina  609295.375   
9                   3306JZR  TRACTORAS  Francisco Herrera Molina  609295.375   
10                  3306JZR  TRACTORAS  Francisco Herrera Molina  609295.375   
...                     ...        ...                       ...         ...   
507803  5497LFN ITER FRUGAL  TRACTORAS  UNKNOWN E28672116V000000   79639.815   
507804  5497LFN ITER FRUGAL  TRACTORAS  UNKNOWN E28672116V000000   79641.310   
507805  5497LFN ITER FRUGAL  TRACTORAS  UNKNOWN E28672116V000000   79642.800   
507806  5497LFN ITER FRUGAL  TRACTORAS  UNKNOWN E28672116V000000   79644.095   
507807  5497LFN ITER FRUGAL  TRACTORAS  UNKNOWN E28672116V000000   79645.550   

        Velocidad  ...  NivelFuel  ConsumoTotal  Frenadas  TiempoRalenti  \
6           0.000  ...       70.4      192041.0    5807.0      5847999.0   
7           0.000  ...       70.4      192041.0    5807.0      5848059.0   
8           0.000  ...       70.4      192041.0    5807.0      5848119.0   
9           0.000  ...       70.4      192041.0    5807.0      5848179.0   
10          0.000  ...       70.4      192041.0    5807.0      5848239.0   
...           ...  ...        ...           ...       ...            ...   
507803     91.748  ...       49.6       25233.0     251.0       589429.0   
507804     92.160  ...       49.2       25233.5     251.0       589429.0   
507805     87.040  ...       49.2       25234.5     251.0       589429.0   
507806     77.616  ...       49.2       25235.5     251.0       589429.0   
507807     92.160  ...       49.2       25236.5     251.0       589429.0   

         ParMotor     Pedal  Embrague        Id  ConductorNum  MatriculaNum  
6       3465938.0  123444.0       NaN         7            27            41  
7       3465938.0  123444.0       NaN         8            27            41  
8       3465938.0  123444.0       NaN         9            27            41  
9       3465938.0  123444.0       NaN        10            27            41  
10      3465938.0  123444.0       NaN        11            27            41  
...           ...       ...       ...       ...           ...           ...  
507803   373818.0    6983.0   59592.0  10729028           157            63  
507804   373818.0    6983.0   59592.0  10729029           157            63  
507805   373831.0    6983.0   59592.0  10729030           157            63  
507806   373890.0    6983.0   59593.0  10729031           157            63  
507807   373915.0    6983.0   59594.0  10729032           157            63  

[10323170 rows x 24 columns]

In [25]:
df.dtypes

Latitud                         float64
Longitud                        float64
Altitud                           int64
Timestamp           datetime64[ns, UTC]
Compania                         object
Matricula                        object
Flota                            object
Conductor                        object
Odometro                        float64
Velocidad                       float64
HorasMotor                      float64
ControlCrucero                  float64
TemperaturaMotor                float64
RPM                             float64
NivelFuel                       float64
ConsumoTotal                    float64
Frenadas                        float64
TiempoRalenti                   float64
ParMotor                        float64
Pedal                           float64
Embrague                        float64
Id                                int32
ConductorNum                      int32
MatriculaNum                      int32
dtype: object

In [26]:
df.isnull().sum()

Latitud                   0
Longitud                  0
Altitud                   0
Timestamp                 0
Compania                  0
Matricula                 0
Flota                     0
Conductor                 0
Odometro               2443
Velocidad                53
HorasMotor               61
ControlCrucero          736
TemperaturaMotor       1303
RPM                     749
NivelFuel             27761
ConsumoTotal         458691
Frenadas             353959
TiempoRalenti           659
ParMotor                892
Pedal                    19
Embrague            1834593
Id                        0
ConductorNum              0
MatriculaNum              0
dtype: int64

Flota y compania, las sacamos del dataset final ya que solo tienen un valor, no tienen variabilidad.

**Imputaciones de NaN**, las que tengan menos de 500.000 las borramos (5% es asumible), resto ya vemos método de imputación, todas excepto embrague.

Odometro               2443

Velocidad               53

HorasMotor              61

ControlCrucero         736

TemperaturaMotor      1303

RPM                   749

NivelFuel             27761

ConsumoTotal         458691

Frenadas             353959

TiempoRalenti          659

ParMotor               892

Pedal                   19

Embrague            1864593 

In [27]:
listaEliminarNa = ['Odometro', 
                   'Velocidad', 
                   'HorasMotor', 
                   'ControlCrucero', 
                   'TemperaturaMotor', 
                   'RPM', 
                   'NivelFuel',
                   'ConsumoTotal',
                   'Frenadas',
                   'TiempoRalenti',
                   'ParMotor',
                   'Pedal'
                  ]

In [28]:
# Bucle para eliminar NaN de columnas de lista anterior
for i in range(len(listaEliminarNa)):
    df = df.dropna(axis=0, subset= [listaEliminarNa[i]])

In [29]:
# Comprobamos eliminación de NaN
df.isnull().sum()

Latitud                   0
Longitud                  0
Altitud                   0
Timestamp                 0
Compania                  0
Matricula                 0
Flota                     0
Conductor                 0
Odometro                  0
Velocidad                 0
HorasMotor                0
ControlCrucero            0
TemperaturaMotor          0
RPM                       0
NivelFuel                 0
ConsumoTotal              0
Frenadas                  0
TiempoRalenti             0
ParMotor                  0
Pedal                     0
Embrague            1653090
Id                        0
ConductorNum              0
MatriculaNum              0
dtype: int64

Nos quedamos con las numéricas y numéricas creadas, descartando las originales, sacamos TimeStamp como datframe, una vez realizada la imputación volvemos a unir por id que coincidan, por esta razón creamos columna **Id** al principio.

In [30]:
dfEnd =df.iloc[:, [0, 1, 2, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]]
dfEnd

Latitud  Longitud  Altitud    Odometro  Velocidad  HorasMotor  \
6       37.390972 -5.939639       21  609295.375      0.000     9668.55   
7       37.390972 -5.939639       21  609295.375      0.000     9668.60   
8       37.390972 -5.939639       21  609295.375      0.000     9668.60   
9       37.390972 -5.939639       21  609295.375      0.000     9668.60   
10      37.390972 -5.939639       21  609295.375      0.000     9668.65   
...           ...       ...      ...         ...        ...         ...   
507803  39.502860 -2.269250      803   79639.815     91.748     1156.40   
507804  39.508390 -2.253667      802   79641.310     92.160     1156.40   
507805  39.513611 -2.238083      831   79642.800     87.040     1156.45   
507806  39.516888 -2.224333      874   79644.095     77.616     1156.45   
507807  39.509418 -2.211278      879   79645.550     92.160     1156.45   

        ControlCrucero  TemperaturaMotor      RPM  NivelFuel  ConsumoTotal  \
6                  0.0              30.0   549.75       70.4      192041.0   
7                  0.0              32.0   549.25       70.4      192041.0   
8                  0.0              33.0   549.00       70.4      192041.0   
9                  0.0              35.0   550.00       70.4      192041.0   
10                 0.0              36.0   548.75       70.4      192041.0   
...                ...               ...      ...        ...           ...   
507803             0.0              78.0  1275.00       49.6       25233.0   
507804             0.0              78.0  1278.00       49.2       25233.5   
507805             0.0              79.0  1210.00       49.2       25234.5   
507806             0.0              84.0  1388.00       49.2       25235.5   
507807             0.0              86.0  1279.00       49.2       25236.5   

        Frenadas  TiempoRalenti   ParMotor     Pedal  Embrague        Id  \
6         5807.0      5847999.0  3465938.0  123444.0       NaN         7   
7         5807.0      5848059.0  3465938.0  123444.0       NaN         8   
8         5807.0      5848119.0  3465938.0  123444.0       NaN         9   
9         5807.0      5848179.0  3465938.0  123444.0       NaN        10   
10        5807.0      5848239.0  3465938.0  123444.0       NaN        11   
...          ...            ...        ...       ...       ...       ...   
507803     251.0       589429.0   373818.0    6983.0   59592.0  10729028   
507804     251.0       589429.0   373818.0    6983.0   59592.0  10729029   
507805     251.0       589429.0   373831.0    6983.0   59592.0  10729030   
507806     251.0       589429.0   373890.0    6983.0   59593.0  10729031   
507807     251.0       589429.0   373915.0    6983.0   59594.0  10729032   

        ConductorNum  MatriculaNum  
6                 27            41  
7                 27            41  
8                 27            41  
9                 27            41  
10                27            41  
...              ...           ...  
507803           157            63  
507804           157            63  
507805           157            63  
507806           157            63  
507807           157            63  

[9507913 rows x 19 columns]

Creamos el dataframe con Timestamp

In [31]:
dfTime = df.iloc[:, [3, 21]]

In [32]:
dfTime

Timestamp        Id
6      2020-01-22 02:57:01+00:00         7
7      2020-01-22 02:58:01+00:00         8
8      2020-01-22 02:59:01+00:00         9
9      2020-01-22 03:00:01+00:00        10
10     2020-01-22 03:01:01+00:00        11
...                          ...       ...
507803 2020-06-22 21:55:46+00:00  10729028
507804 2020-06-22 21:56:46+00:00  10729029
507805 2020-06-22 21:57:46+00:00  10729030
507806 2020-06-22 21:58:46+00:00  10729031
507807 2020-06-22 21:59:46+00:00  10729032

[9507913 rows x 2 columns]

Vamos a imputar los Nan de **Embrague**, interpolando con datos conocidos, podemos verlo en el siguienyte enlace:  
http://pyciencia.blogspot.com/2015/04/trabajar-con-datos-nan-en-dataframe.html   

In [33]:
dfEnd = dfEnd.interpolate(method='linear')

In [34]:
dfEnd.isnull().sum()

Latitud                 0
Longitud                0
Altitud                 0
Odometro                0
Velocidad               0
HorasMotor              0
ControlCrucero          0
TemperaturaMotor        0
RPM                     0
NivelFuel               0
ConsumoTotal            0
Frenadas                0
TiempoRalenti           0
ParMotor                0
Pedal                   0
Embrague            11542
Id                      0
ConductorNum            0
MatriculaNum            0
dtype: int64

Realizada la imputación aún queda 11542 que los eliminaremos por ser una cifra muy baja respecto al total de registros.

In [35]:
dfEnd.dropna()

Latitud  Longitud  Altitud    Odometro  Velocidad  HorasMotor  \
11811   37.390473 -5.939722       21  759022.915      0.000    12663.10   
11812   37.390473 -5.939722       21  759022.915      0.000    12663.10   
11813   37.390473 -5.939722       21  759022.915      0.000    12663.10   
11814   37.390473 -5.939722       21  759022.925      3.384    12663.15   
11815   37.390473 -5.939722       21  759022.940      0.000    12663.15   
...           ...       ...      ...         ...        ...         ...   
507803  39.502860 -2.269250      803   79639.815     91.748     1156.40   
507804  39.508390 -2.253667      802   79641.310     92.160     1156.40   
507805  39.513611 -2.238083      831   79642.800     87.040     1156.45   
507806  39.516888 -2.224333      874   79644.095     77.616     1156.45   
507807  39.509418 -2.211278      879   79645.550     92.160     1156.45   

        ControlCrucero  TemperaturaMotor     RPM  NivelFuel  ConsumoTotal  \
11811              0.0              12.0   605.5       40.0      261869.0   
11812              0.0              14.0   597.5       40.0      261869.0   
11813              0.0              16.0   599.0       40.0      261869.5   
11814              0.0              18.0   598.0       40.0      261869.5   
11815              0.0              20.0   600.0       40.0      261869.5   
...                ...               ...     ...        ...           ...   
507803             0.0              78.0  1275.0       49.6       25233.0   
507804             0.0              78.0  1278.0       49.2       25233.5   
507805             0.0              79.0  1210.0       49.2       25234.5   
507806             0.0              84.0  1388.0       49.2       25235.5   
507807             0.0              86.0  1279.0       49.2       25236.5   

        Frenadas  TiempoRalenti  ParMotor     Pedal  Embrague        Id  \
11811     9614.0      6019315.0  483817.0  130083.0  385986.0     11812   
11812     9614.0      6019375.0  483817.0  130083.0  385986.0     11813   
11813     9614.0      6019435.0  483817.0  130083.0  385986.0     11814   
11814     9614.0      6019484.0  483817.0  130083.0  385988.0     11815   
11815     9614.0      6019528.0  483817.0  130083.0  385989.0     11816   
...          ...            ...       ...       ...       ...       ...   
507803     251.0       589429.0  373818.0    6983.0   59592.0  10729028   
507804     251.0       589429.0  373818.0    6983.0   59592.0  10729029   
507805     251.0       589429.0  373831.0    6983.0   59592.0  10729030   
507806     251.0       589429.0  373890.0    6983.0   59593.0  10729031   
507807     251.0       589429.0  373915.0    6983.0   59594.0  10729032   

        ConductorNum  MatriculaNum  
11811            189           106  
11812            189           106  
11813            189           106  
11814            189           106  
11815            189           106  
...              ...           ...  
507803           157            63  
507804           157            63  
507805           157            63  
507806           157            63  
507807           157            63  

[9496371 rows x 19 columns]

In [36]:
dfEnd.isnull().sum()

Latitud                 0
Longitud                0
Altitud                 0
Odometro                0
Velocidad               0
HorasMotor              0
ControlCrucero          0
TemperaturaMotor        0
RPM                     0
NivelFuel               0
ConsumoTotal            0
Frenadas                0
TiempoRalenti           0
ParMotor                0
Pedal                   0
Embrague            11542
Id                      0
ConductorNum            0
MatriculaNum            0
dtype: int64

No las ha borrado, probamos con otro método.

In [37]:
# Eliminamos filas con NAN, en Embrague que es en la única que quedan
dfEnd = dfEnd[~np.isnan(dfEnd).any(axis=1)]

In [38]:
dfEnd.isnull().sum()

Latitud             0
Longitud            0
Altitud             0
Odometro            0
Velocidad           0
HorasMotor          0
ControlCrucero      0
TemperaturaMotor    0
RPM                 0
NivelFuel           0
ConsumoTotal        0
Frenadas            0
TiempoRalenti       0
ParMotor            0
Pedal               0
Embrague            0
Id                  0
ConductorNum        0
MatriculaNum        0
dtype: int64

In [39]:
dfTime.isnull().sum()

Timestamp    0
Id           0
dtype: int64

Vamos a pasar la mayoría de variables a tipo **int** ya que float ocupa mucha memoria y da problemas de memoria valga la redundancia. Dejamos **ConsumoTotal** que va realizando anotaciones cada 0.5, y como será nuestra variable objetivo pues para que sea más fidedigna, también dejamos como float **Odometro**..

In [40]:
columnas = ['Velocidad', 'TemperaturaMotor', 'Frenadas', 'TiempoRalenti', 'ParMotor', 'Pedal', 'Embrague', 'RPM', 
            'ControlCrucero', 'NivelFuel', 'HorasMotor']

In [41]:
for c in columnas:
    dfEnd[c] = dfEnd[c].astype(int)

In [42]:
dfEnd.dtypes

Latitud             float64
Longitud            float64
Altitud               int64
Odometro            float64
Velocidad             int32
HorasMotor            int32
ControlCrucero        int32
TemperaturaMotor      int32
RPM                   int32
NivelFuel             int32
ConsumoTotal        float64
Frenadas              int32
TiempoRalenti         int32
ParMotor              int32
Pedal                 int32
Embrague              int32
Id                    int32
ConductorNum          int32
MatriculaNum          int32
dtype: object

In [43]:
finalDF = pd.merge(dfEnd, dfTime)

In [44]:
dfEnd.shape

(9496371, 19)

In [45]:
dfTime.shape

(9507913, 2)

In [46]:
finalDF

Latitud  Longitud  Altitud    Odometro  Velocidad  HorasMotor  \
0        37.390473 -5.939722       21  759022.915          0       12663   
1        37.390473 -5.939722       21  759022.915          0       12663   
2        37.390473 -5.939722       21  759022.915          0       12663   
3        37.390473 -5.939722       21  759022.925          3       12663   
4        37.390473 -5.939722       21  759022.940          0       12663   
...            ...       ...      ...         ...        ...         ...   
9496366  39.502860 -2.269250      803   79639.815         91        1156   
9496367  39.508390 -2.253667      802   79641.310         92        1156   
9496368  39.513611 -2.238083      831   79642.800         87        1156   
9496369  39.516888 -2.224333      874   79644.095         77        1156   
9496370  39.509418 -2.211278      879   79645.550         92        1156   

         ControlCrucero  TemperaturaMotor   RPM  NivelFuel  ConsumoTotal  \
0                     0                12   605         40      261869.0   
1                     0                14   597         40      261869.0   
2                     0                16   599         40      261869.5   
3                     0                18   598         40      261869.5   
4                     0                20   600         40      261869.5   
...                 ...               ...   ...        ...           ...   
9496366               0                78  1275         49       25233.0   
9496367               0                78  1278         49       25233.5   
9496368               0                79  1210         49       25234.5   
9496369               0                84  1388         49       25235.5   
9496370               0                86  1279         49       25236.5   

         Frenadas  TiempoRalenti  ParMotor   Pedal  Embrague        Id  \
0            9614        6019315    483817  130083    385986     11812   
1            9614        6019375    483817  130083    385986     11813   
2            9614        6019435    483817  130083    385986     11814   
3            9614        6019484    483817  130083    385988     11815   
4            9614        6019528    483817  130083    385989     11816   
...           ...            ...       ...     ...       ...       ...   
9496366       251         589429    373818    6983     59592  10729028   
9496367       251         589429    373818    6983     59592  10729029   
9496368       251         589429    373831    6983     59592  10729030   
9496369       251         589429    373890    6983     59593  10729031   
9496370       251         589429    373915    6983     59594  10729032   

         ConductorNum  MatriculaNum                 Timestamp  
0                 189           106 2020-01-22 11:33:27+00:00  
1                 189           106 2020-01-22 11:34:27+00:00  
2                 189           106 2020-01-22 11:35:27+00:00  
3                 189           106 2020-01-22 11:36:27+00:00  
4                 189           106 2020-01-22 11:37:27+00:00  
...               ...           ...                       ...  
9496366           157            63 2020-06-22 21:55:46+00:00  
9496367           157            63 2020-06-22 21:56:46+00:00  
9496368           157            63 2020-06-22 21:57:46+00:00  
9496369           157            63 2020-06-22 21:58:46+00:00  
9496370           157            63 2020-06-22 21:59:46+00:00  

[9496371 rows x 20 columns]

In [47]:
# Eliminamos columna Id que ya no la necesitamos
finalDF = finalDF.drop(['Id'], axis = 1)

In [48]:
finalDF

Latitud  Longitud  Altitud    Odometro  Velocidad  HorasMotor  \
0        37.390473 -5.939722       21  759022.915          0       12663   
1        37.390473 -5.939722       21  759022.915          0       12663   
2        37.390473 -5.939722       21  759022.915          0       12663   
3        37.390473 -5.939722       21  759022.925          3       12663   
4        37.390473 -5.939722       21  759022.940          0       12663   
...            ...       ...      ...         ...        ...         ...   
9496366  39.502860 -2.269250      803   79639.815         91        1156   
9496367  39.508390 -2.253667      802   79641.310         92        1156   
9496368  39.513611 -2.238083      831   79642.800         87        1156   
9496369  39.516888 -2.224333      874   79644.095         77        1156   
9496370  39.509418 -2.211278      879   79645.550         92        1156   

         ControlCrucero  TemperaturaMotor   RPM  NivelFuel  ConsumoTotal  \
0                     0                12   605         40      261869.0   
1                     0                14   597         40      261869.0   
2                     0                16   599         40      261869.5   
3                     0                18   598         40      261869.5   
4                     0                20   600         40      261869.5   
...                 ...               ...   ...        ...           ...   
9496366               0                78  1275         49       25233.0   
9496367               0                78  1278         49       25233.5   
9496368               0                79  1210         49       25234.5   
9496369               0                84  1388         49       25235.5   
9496370               0                86  1279         49       25236.5   

         Frenadas  TiempoRalenti  ParMotor   Pedal  Embrague  ConductorNum  \
0            9614        6019315    483817  130083    385986           189   
1            9614        6019375    483817  130083    385986           189   
2            9614        6019435    483817  130083    385986           189   
3            9614        6019484    483817  130083    385988           189   
4            9614        6019528    483817  130083    385989           189   
...           ...            ...       ...     ...       ...           ...   
9496366       251         589429    373818    6983     59592           157   
9496367       251         589429    373818    6983     59592           157   
9496368       251         589429    373831    6983     59592           157   
9496369       251         589429    373890    6983     59593           157   
9496370       251         589429    373915    6983     59594           157   

         MatriculaNum                 Timestamp  
0                 106 2020-01-22 11:33:27+00:00  
1                 106 2020-01-22 11:34:27+00:00  
2                 106 2020-01-22 11:35:27+00:00  
3                 106 2020-01-22 11:36:27+00:00  
4                 106 2020-01-22 11:37:27+00:00  
...               ...                       ...  
9496366            63 2020-06-22 21:55:46+00:00  
9496367            63 2020-06-22 21:56:46+00:00  
9496368            63 2020-06-22 21:57:46+00:00  
9496369            63 2020-06-22 21:58:46+00:00  
9496370            63 2020-06-22 21:59:46+00:00  

[9496371 rows x 19 columns]

El set original partía con 10729032 registros y nos quedamos con 9496371 , hemos eliminado un 11 % que es asumible.
Guardamos en csv para probar en Spark

In [49]:
#finalDF.to_csv ('wemob.csv')

9 minutos tarda en ejecutarse todo el código en mi equipo.